pennylane/templates/subroutines/hilbert_schmidt.py

```diff
@@ -0,0 +1,256 @@
+# Copyright 2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This submodule contains the templates for Hilbert Schmidt tests.
+"""
+# pylint: disable-msg=too-many-arguments
+import pennylane as qml
+from pennylane.operation import AnyWires, Operation
+
+
+class HilbertSchmidt(Operation):
+    r"""Create a Hilbert Schmidt template that can be used to compute the Hilbert Schmidt Test (HST). The HST is a
+    useful quantity used when we want to compile an unitary `U` with an approximate unitary `V`. The HST is used as a
+    distance between `U` and `V`, the value of the HST is 0 if and only if `V` is equal to `U` (up to global phase).
+    Therefore we can define a cost by:
+
+    .. math::
+        C_{HST} = 1 - \frac{1}{d^2} \left|Tr(V^{\dagger}U)\right|^2,
+
+    where the quantity :math:`\frac{1}{d^2} \left|Tr(V^{\dagger}U)\right|^2` is the Hilbert Schmidt Test. It is
+    equivalent to taking the outcome probability of the state :math:`|0 ... 0\rangle` for the following circuit:
+
+    .. figure:: ../../_static/templates/subroutines/hst.png
+        :align: center
+        :width: 60%
+        :target: javascript:void(0);
+
+    It defines our decomposition for the Hilbert Schmidt Test template.
+
+    Args:
+        params (array): Parameters for the quantum function `V`.
+        v_function (Callable): Quantum function that represents the approximate compiled unitary `V`.
+        v_wires (int or Iterable[Number, str]]): the wire(s) the approximate compiled unitary act on.
+        u_tape (.QuantumTape): `U`, the unitary to be compiled as a ``qml.tape.QuantumTape``.
+
+    Raises:
+        QuantumFunctionError: The argument u_tape must be a QuantumTape.
```

Comment:
Minor: maybe having `QuantumFunctionError` once and having a list of the outcomes could work too.

Reply:
Something like ['String1', 'String2', ...]?

Reply:
Oh more like:
```
        QuantumFunctionError:

                * The argument u_tape must be a QuantumTape.
                * ``v_function`` is not a valid quantum function.
                * ``U`` and ``V`` do not have the same number of wires.
```
I don't think we have any conventions for this case though, so just a thought.